File: examples/concurrency/concurrency.rb
require 'thwait'
require 'benchmark'

busy_work = Proc.new do |id|
  iter_count = 300 + Random.rand(100)

  time = Benchmark.measure do
    iter_count.times do |i|
      iter_count.times do |j|
        mul = i * j

        File.open("/dev/null", "w") { |file|
          file.puts mul
        }
      end
    end
  end

  puts "Hello, world! from #{id} busy_work; Iter Count: #{iter_count}\n Took: #{time}"
end

# busy_work.call(1)

# -------- Threads ----------

threads = 4.times.map do |id|
 Thread.new do
  busy_work.call(id)
 end
end

ThWait.all_waits(threads)

# -------- Processes --------

# 4.times do |id|
#   pid = Process.fork do
#     busy_work.call(id)
#   end
#   puts pid
# end

# Process.wait

File: examples/concurrency/concurrency.go
package main

import (
	"bufio"
	"crypto/rand"
	"fmt"
	"io/ioutil"
	"math/big"
	"os"
	"strconv"
)

func busyWork(iterCount int) {
	for i := 0; i < iterCount; i++ {
		for j := 0; j < iterCount; j++ {
			mul := []byte(strconv.Itoa(i * j))
			ioutil.WriteFile("/dev/null", mul, 0644)
		}
	}
}

func getRandomIterCount() int {
	randomFile, _ := os.Open("/dev/random")
	randomReader := bufio.NewReader(randomFile)
	randomNumber, _ := rand.Int(randomReader, big.NewInt(400))
	iterCount := int(100 + randomNumber.Int64())

	return iterCount
}

func say(s string, index int) {
	iterCount := getRandomIterCount()

	busyWork(iterCount)

	fmt.Printf("Hello, %s! from %d goroutine; Iter Count: %d\n", s, index+1, iterCount)
}

func main() {
	for i := 0; i < 4; i++ {
		say("world", i)
	}

	// time.Sleep(10 * time.Second)
}
